### ISSUE TITLE
Not sure if issue: queryOnly devices can be controlled

### ISSUE THREAD

**Issue by NikAWing** [mikejac/node-red-contrib-google-smarthome#404]
Hey all!

I'm not sure if this is related to the smarthome node, a google bug or something generated by a dumb user ;)

I added 2 sensors, light and door and set both to queryOnly On/Off resp. Open/Close.
This works, but in the google home app, if I click on the buttons there, it is not read-only, it immediately changes the state shown below the smarthome device nodes in NodeRed.

Is this a google problem or what can I do to just DISPLAY the state of the door and light/lux sensor?
In this case I only need to know if it's open/closed or on/off but without being able to set something in the app.

**Comment by Caprico85** [comment]
Hello @NikAWing 

How are your devices configured? I have a window set as query only.

![image](https://github.com/mikejac/node-red-contrib-google-smarthome/assets/2081806/2551673d-36ce-4c47-8770-8e6fa2d65c36)


It correctly shows up as read only in the app.

![image](https://github.com/mikejac/node-red-contrib-google-smarthome/assets/2081806/2216d623-ef62-4463-b054-efa930c36ceb)

**Comment by NikAWing** [comment]
for me it's kinda the same:
![grafik](https://github.com/mikejac/node-red-contrib-google-smarthome/assets/76008757/048f3176-8534-49d6-a35c-e522b31154a1)

But in the home app I can click on the buttons and see the reaction in node red.
Or should I use sensor instead?
It puzzles me that "query only" does not have "read only" attributes :)

**Comment by Caprico85** [comment]
You are right. My first test was with the open/close trait. There it works. But with the on/off trait, query only is ignored and you can control the lights. I'll see if I can fix this.

**Comment by Caprico85** [comment]
The parameters we send seems to be correct. But the whole logic around query/command only seems to be seriously broken by Google.

I experimented with several different traits, in the app or via voice, with querying and commanding. And I found no single trait that works all of the time.

Sometimes you can just query or control as if no "query/command only" was set. Sometimes actions are correctly denied, but with totally unrelated error messages like "This device is not set up". Sometimes you see the readonly icon in the app, sometimes you don't. And there's not even a recognizable pattern on what works and what does not.

So currently I am not able to fix this. I think I'll have to write a ticket on Google's official bugtracker.

**Comment by NikAWing** [comment]
Ok so it IS google related, thanks for checking! :)

I have also played around with devices and options (because I don't really find the simple ones I want to use)
And yeah, the app also reacts weird, you can switch a light the 1st time to touch the button, the second time you touch it opens another window with a bigger switch and this also suddenly reacts to brightness levels etc which were not enabled in the node.

Currently it is (and I'm still a beginner with google smarthome) not easy for me to setup simple logic.
Why don't they have an easy toggle/push button switch for example garage door opener, it is just a push button that triggers [direction1-stop-direction2-stop] or a simple 3state switch as in 1-off-2, so when you click on 2, 1 goes off, if you click on 1, 2 goes off and if you click on off, both are off.

I also noticed updating the switch nodes with "on:true/false" sometimes works and you see it in the app without reloading and for other switches of the same type and settings updating does not work until you refresh the app.
But this makes it look that all problems are google related.

BTW, do you know any group/community where you can ask for help on how to set up things I just wrote about?
I don't think opening an issue thread here for things like that is a good idea :)

**Comment by Caprico85** [comment]
I opened an issue on Google's bugtracker regarding the "query only" issue: https://issuetracker.google.com/issues/296484985
Not sure if it helps though. The problem seems to exist for at least 3 years now: https://stackoverflow.com/questions/61955823/how-to-tell-google-to-not-send-execute-intents-for-sensor

&nbsp;
I'm having the same problems in the app. Mostly, I don't even send the current state of my devices into the app and use the Node-RED dashboard instead.

&nbsp;
> BTW, do you know any group/community where you can ask for help on how to set up things I just wrote about?

We have discussions enabled now for our repository, thanks to master @mikejac :-)

**Comment by Caprico85** [comment]
Got a response from Google:

>Dear Developer,
>Thank you for reaching out to us. Based on the details you've provided, it seems that the root cause of the problem lies in the implementation of error handling. Specifically, your backend server needs to respond with the error code functionNotSupported when the queryOnlyOnOff is set to true.
> To illustrate, let's consider the following scenario. Suppose the user sends a command to "turn off schreibtischlampe":
> {
>  "requestId": "6314316342930951374",
>  "payload": {
>    "commands": [
>      {
>        "ids": [
>          "475525b40c0b1b0f"
>        ],
>        "status": "ERROR",
>        "errorCode": "functionNotSupported"
>      }
>    ]
>  }
> }

We really don't send this error message. I'll fix that.

But I think this won't fix the whole problem. Sending this error will maybe fix the "device not set up" voice message when trying to control a read only device. But I guess it won't fix the buttons in the app. The buttons are already shown before we have a chance to send this `functionNotSupported` response.

**Comment by Caprico85** [comment]
Still discussing with Google. This is reproducible on the Google Smart Home Playground. So the problem really is on Google's side and not a wrong implementation in our code.

**Comment by NikAWing** [comment]
that is great to know :) so the chances for an improvement or fix are good if the google staff looks further into it.

I've seen you've also enabled the github discussions area, that's great!
I currently have to some IRL hardships to overcome and not much time, I will definitely join later and ask questions :)

**Comment by Caprico85** [comment]
We're getting closer step by step...

Google answered that they are not able to reproduce this. So I did some more testing and noticed that this happens if Google Assistant is set to German or Dutch, but not if it is set to english. I hope, Google is now finally able to reproduce this.

Viele Grüße 😄

**Comment by ckhmer1** [comment]
Great news, thanks. It also happens in Italian.

**Comment by Caprico85** [comment]
Google just marked my bug report on their tracker (https://issuetracker.google.com/issues/296484985) as fixed. Had no time to try it yet.

**Comment by NikAWing** [comment]
I'm still on 0.4.9 since it worked kinda great.
I need to upgrade to newest version and have to read into the app.js upload first, lol

**Comment by NikAWing** [comment]
nevermind, I just updated the node in node-red and it works as before
also "query-only" still shows the same behavior as before, hmmm
